# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

# Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [9]:
from csv import reader
# open file in read mode
with open('rural_population_percent.csv', 'r') as read_obj:
    # pass the file object to reader() to get the reader object
    csv_reader = reader(read_obj)
    # Iterate over each row in the csv using reader object
    for row in csv_reader:
        # row variable is a list that represents a row in csv
        print(row)

['\ufeff"Data Source"', 'World Development Indicators', '']
[]
['Last Updated Date', '2018-06-28', '']
[]
['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code', '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968', '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977', '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986', '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995', '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017', '']
['Aruba', 'ABW', 'Rural population (% of total population)', 'SP.RUR.TOTL.ZS', '49.224', '49.239', '49.254', '49.27', '49.285', '49.3', '49.315', '49.33', '49.346', '49.361', '49.376', '49.391', '49.407', '49.422', '49.437', '49.452', '49.468', '49.483', '49.498', '49.513', '49.528', '49.544', '49.559', '49.574', '49.589', '49.605', '49.62', '49.635', '49

In [10]:
from csv import reader
# open file in read mode
with open('electricity_access_percent.csv', 'r') as read_obj:
    # pass the file object to reader() to get the reader object
    csv_reader = reader(read_obj)
    # Iterate over each row in the csv using reader object
    for row in csv_reader:
        # row variable is a list that represents a row in csv
        print(row)

['\ufeff"Data Source"', 'World Development Indicators', '']
[]
['Last Updated Date', '2018-06-28', '']
[]
['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code', '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968', '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977', '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986', '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995', '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017', '']
['Aruba', 'ABW', 'Access to electricity (% of population)', 'EG.ELC.ACCS.ZS', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '88.4453506469727', '88.7808456420898', '89.1158294677734', '89.44775390625', '89.7735595703125', '90.0901870727539', '90.3945846557617', '90.6836776733398', '

In [11]:
# TODO: import the pandas library
import pandas as pd
# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
# HINT: The file paths are 'rural_population_percent.csv' and 'electricity_access_percent.csv'
df_rural = pd.read_csv("rural_population_percent.csv",skiprows=4)
df_electricity = pd.read_csv("electricity_access_percent.csv",skiprows=4)

# TODO: remove the 'Unnamed:62' column from each data set
df_rural.drop(['Unnamed: 62'],axis = 1,inplace=True)
df_electricity.drop(['Unnamed: 62'],axis = 1,inplace=True)
# TODO: combine the two data sets together using the concat method
# In other words, all of the rows of df_rural will come first
# followed by all the rows in df_electricity. This is possible to do
# because they both have the same column names.
df_compined = pd.concat([df_rural,df_electricity])

In [12]:
df_compined.shape

(528, 62)

In [13]:
df_compined.head()

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   
2       Angola          AGO  Rural population (% of total population)   
3      Albania          ALB  Rural population (% of total population)   
4      Andorra          AND  Rural population (% of total population)   

   Indicator Code    1960    1961    1962    1963    1964    1965   ...    \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.270  49.285  49.300   ...     
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.890  90.574  90.250   ...     
2  SP.RUR.TOTL.ZS  89.565  89.202  88.796  88.376  87.942  87.496   ...     
3  SP.RUR.TOTL.ZS  69.295  69.057  68.985  68.914  68.842  68.770   ...     
4  SP.RUR.TOTL.ZS  41.550  39.017  36.538  34.128  31.795  29.555   ...     

     2008    2009    2010    2011    2012    2013    2014    2015    2016  \
0  56.217  56.579  56.941  57.302  57.636  57.942  58.221  58.472  58.696   
1  76.054  75.687  75.311  74.926  74.532  74.129  73.718  73.297  72.868   
2  61.491  60.701  59.903  59.100  58.301  57.510  56.726  55.950  55.181   
3  50.009  48.924  47.837  46.753  45.670  44.617  43.591  42.593  41.624   
4  11.133  11.648  12.183  12.740  13.292  13.835  14.367  14.885  15.388   

     2017  
0  58.893  
1  72.430  
2  54.422  
3  40.684  
4  15.873  

[5 rows x 62 columns]

# Exercise 2 (Challenge)

This exercise is more challenging.

Combine the two datas in the csv file together so that the output looks like the following:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [14]:
df_rural.columns

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017'],
      dtype='object')

In [15]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value
df_rural_melted = pd.melt(df_rural,id_vars=['Country Name','Country Code','Indicator Name','Indicator Code'],\
                          var_name = 'Year' , value_name='Electricity Value')

df_electricity_melted = pd.melt(df_electricity,id_vars=['Country Name','Country Code','Indicator Name','Indicator Code'],\
                          var_name = 'Year' , value_name='Rural Value')

# TODO: drop any columns from the data frames that aren't needed
df_rural_melted.drop(['Indicator Name','Indicator Code'],axis=1,inplace = True)
df_electricity_melted.drop(['Indicator Name','Indicator Code'],axis=1,inplace = True)

# TODO: merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year
df_merged = pd.merge(df_rural_melted,df_electricity_melted, how='outer', on=["Country Name", "Country Code","Year"])
# TODO: sort the results by country and then by year

df_combined = df_merged.sort_values(["Country Name","Year"])
df_combined.head()

Country Name Country Code  Year  Electricity Value  Rural Value
1     Afghanistan          AFG  1960             91.779          NaN
265   Afghanistan          AFG  1961             91.492          NaN
529   Afghanistan          AFG  1962             91.195          NaN
793   Afghanistan          AFG  1963             90.890          NaN
1057  Afghanistan          AFG  1964             90.574          NaN

In [16]:
df_combined.head()

Country Name Country Code  Year  Electricity Value  Rural Value
1     Afghanistan          AFG  1960             91.779          NaN
265   Afghanistan          AFG  1961             91.492          NaN
529   Afghanistan          AFG  1962             91.195          NaN
793   Afghanistan          AFG  1963             90.890          NaN
1057  Afghanistan          AFG  1964             90.574          NaN